File: NewsDemo/app/src/main/java/com/kevin/newsdemo/base/BaseResult.java
package com.kevin.newsdemo.base;

import java.util.Objects;

/**
 * Created by kevin on 2019/07/17 13:17.
 */
public class BaseResult<T> {
    ResultCode code;
    T t;
    Throwable throwable;

    private BaseResult(ResultCode code, T t) {
        this.code = code;
        this.t = t;
    }

    private BaseResult(ResultCode code, Throwable throwable) {
        this.code = code;
        this.throwable = throwable;
    }

    public static <T> BaseResult<T> succeed(T t) {
        return new BaseResult(ResultCode.OK, t);
    }

    public static BaseResult error(Throwable t) {
        return new BaseResult(ResultCode.ERROR, t);
    }

    public static BaseResult error(ResultCode code, Throwable t) {
        return new BaseResult(code, t);
    }

    public boolean isOK() {
        return code == ResultCode.OK;
    }

    public T getData() {
        return t;
    }

    public ResultCode getCode() {
        return code;
    }

    public Throwable getThrowable() {
        return throwable;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o)
            return true;
        if (!(o instanceof BaseResult))
            return false;
        BaseResult<?> that = (BaseResult<?>) o;
        return code == that.code &&
          Objects.equals(t, that.t) &&
          Objects.equals(throwable, that.throwable);
    }

    @Override
    public int hashCode() {
        return Objects.hash(code, t, throwable);
    }
}

File: NewsDemo/app/src/main/java/com/kevin/newsdemo/base/BaseResult.kt
package com.kevin.newsdemo.base

import java.util.*

/**
 * Created by kevin on 2019/07/17 13:17.
 */
class BaseResult<T>(var code: ResultCode, var data:T?, var throwable: Throwable?) {
    val isOK: Boolean
        get() = code === ResultCode.OK

    private constructor(code: ResultCode, t: T) : this(code,t,null) {
        this.code = code
        this.data = t
    }

    private constructor(code: ResultCode, throwable: Throwable) :this(code,null,throwable) {
        this.code = code
        this.throwable = throwable
    }

    override fun equals(o: Any?): Boolean {
        if (this === o)
            return true
        if (o !is BaseResult<*>)
            return false
        val that = o as BaseResult<*>?
        return code === that!!.code &&
                data == that!!.data &&
                throwable == that.throwable
    }

    override fun hashCode(): Int {
        return Objects.hash(code, data, throwable)
    }

    companion object {

        fun <T> succeed(t: T): BaseResult<T> {
            return BaseResult(ResultCode.OK, t)
        }

        fun error(t: Throwable): BaseResult<Any> {
            return BaseResult(ResultCode.ERROR, t)
        }

        fun error(code: ResultCode, t: Throwable): BaseResult<Any> {
            return BaseResult(code, t)
        }
    }
}
